java/vector/src/test/java/org/apache/arrow/vector/TestListViewVector.java

```diff
@@ -0,0 +1,1599 @@
+      innerList1.addOrGetVector(FieldType.nullable(listType.getType()));
+
+      ListViewVector innerList2 = (ListViewVector) innerList1.getDataVector();
+      innerList2.addOrGetVector(FieldType.nullable(listType.getType()));
+
+      ListViewVector innerList3 = (ListViewVector) innerList2.getDataVector();
+      innerList3.addOrGetVector(FieldType.nullable(listType.getType()));
+
+      ListViewVector innerList4 = (ListViewVector) innerList3.getDataVector();
+      innerList4.addOrGetVector(FieldType.nullable(listType.getType()));
+
+      ListViewVector innerList5 = (ListViewVector) innerList4.getDataVector();
+      innerList5.addOrGetVector(FieldType.nullable(listType.getType()));
+
+      ListViewVector innerList6 = (ListViewVector) innerList5.getDataVector();
+      innerList6.addOrGetVector(FieldType.nullable(scalarType.getType()));
+
+      listViewVector.setInitialCapacity(128);
+
+      validateVector(listViewVector);
+    }
+  }
+
+  private void setValuesInBuffer(int[] bufValues, ArrowBuf buffer, long bufWidth) {
+    for (int i = 0; i < bufValues.length; i++) {
+      buffer.setInt(i * bufWidth, bufValues[i]);
+    }
+  }
+
+  /*
+   * Setting up the buffers directly needs to be validated with the base method used in
+   * the ListVector class where we use the approach of startList(),
+   * write to the child vector and endList().
+   * <p>
+   * To support this, we have to consider the following scenarios;
+   * <p>
+   * 1. Only using directly buffer-based inserts.
+   * 2. Default list insertion followed by buffer-based inserts.
+   * 3. Buffer-based inserts followed by default list insertion.
+   */
+  @Test
+  public void testBasicListViewSet() {
+
+    try (ArrowBuf newOffSetBuf = allocator.buffer(1024);
+        ArrowBuf newSizeBuffer = allocator.buffer(1024);
+        ArrowBuf validityBuffer = allocator.buffer(DataSizeRoundingUtil.divideBy8Ceil(1024));
+        BigIntVector childVector = new BigIntVector("element-vector", allocator);
+        ListViewVector listViewVector = ListViewVector.empty("sourceVector", allocator)) {
+      listViewVector.allocateNew();
+
+      childVector.allocateNew(7);
+
+      childVector.set(0, 12);
+      childVector.set(1, -7);
+      childVector.set(2, 25);
+      childVector.set(3, 0);
+      childVector.set(4, -127);
+      childVector.set(5, 127);
+      childVector.set(6, 50);
+
+      childVector.setValueCount(7);
+
+      int[] offSetValues = new int[]{0, 3, 3, 7};
+      int[] sizeValues = new int[]{3, 0, 4, 0};
+
+      BitVectorHelper.setBit(validityBuffer, 0);
+      BitVectorHelper.setBit(validityBuffer, 2);
+      BitVectorHelper.setBit(validityBuffer, 3);
+
+      setValuesInBuffer(offSetValues, newOffSetBuf, BaseRepeatedValueViewVector.OFFSET_WIDTH);
+      setValuesInBuffer(sizeValues, newSizeBuffer, BaseRepeatedValueViewVector.SIZE_WIDTH);
+
+      listViewVector.set(newOffSetBuf, newSizeBuffer, validityBuffer, childVector, 4);
```

Comment:
Er, this API is entirely inconsistent with how regular vectors work. We don't have APIs that let you just arbitrarily override a buffer or child.

Reply:
Hmm... 

```bash
listViewVector.setOffSet(/*index*/, */offset*/)
listViewVector.setSize(/*index*/, */offset*/)
listViewVector.setChildVector(/*childVector*/))
```

Would something like this be okay?

Reply:
There's no need to have setChildVector.

Reply:
So how would the use insert list items? Use the `UnionListViewWriter`?

Reply:
You can directly getDataVector and append to the child. You can use initializeChildrenFromFields to force the child to be a certain type.

Reply:
Let me take a look. Thanks David.

Reply:
Is this acceptable? 

```java
/*
  * Setting up buffers directly would require the following steps to be taken
  * 1. Set offset and size buffers using `setOffSet` and `setSize` methods.
  * 2. Set validity buffer using `setValidity` method.
  * 3. Set lastSet value using `setLastSet` method.
  * 4. Initialize the child vector using `initializeChildrenFromFields` method.
  * 5. Set values in the child vector.
  */
  @Test
  public void testBasicListViewSet1() {

    try (ListViewVector listViewVector = ListViewVector.empty("sourceVector", allocator)) {
      listViewVector.allocateNew();

      listViewVector.setOffSet(0, 0);
      listViewVector.setOffSet(1, 3);
      listViewVector.setOffSet(2, 3);
      listViewVector.setOffSet(3, 7);

      listViewVector.setSize(0, 3);
      listViewVector.setSize(1, 0);
      listViewVector.setSize(2, 4);
      listViewVector.setSize(3, 0);

      listViewVector.setValidity(0, 1);
      listViewVector.setValidity(1, 0);
      listViewVector.setValidity(2, 1);
      listViewVector.setValidity(3, 1);

      listViewVector.setLastSet(3);

      FieldType fieldType = new FieldType(true, new ArrowType.Int(64, true),
          null, null);
      Field field = new Field("element-vector", fieldType, null);
      listViewVector.initializeChildrenFromFields(Collections.singletonList(field));

      FieldVector fieldVector = listViewVector.getDataVector();
      fieldVector.clear();

      BigIntVector childVector = (BigIntVector) fieldVector;
      childVector.allocateNew(7);

      childVector.set(0, 12);
      childVector.set(1, -7);
      childVector.set(2, 25);
      childVector.set(3, 0);
      childVector.set(4, -127);
      childVector.set(5, 127);
      childVector.set(6, 50);

      childVector.setValueCount(7);

      final ArrowBuf offSetBuffer = listViewVector.getOffsetBuffer();
      final ArrowBuf sizeBuffer = listViewVector.getSizeBuffer();

      // check offset buffer
      assertEquals(0, offSetBuffer.getInt(0 * BaseRepeatedValueViewVector.OFFSET_WIDTH));
      assertEquals(3, offSetBuffer.getInt(1 * BaseRepeatedValueViewVector.OFFSET_WIDTH));
      assertEquals(3, offSetBuffer.getInt(2 * BaseRepeatedValueViewVector.OFFSET_WIDTH));
      assertEquals(7, offSetBuffer.getInt(3 * BaseRepeatedValueViewVector.OFFSET_WIDTH));

      // check size buffer
      assertEquals(3, sizeBuffer.getInt(0 * BaseRepeatedValueViewVector.SIZE_WIDTH));
      assertEquals(0, sizeBuffer.getInt(1 * BaseRepeatedValueViewVector.SIZE_WIDTH));
      assertEquals(4, sizeBuffer.getInt(2 * BaseRepeatedValueViewVector.SIZE_WIDTH));
      assertEquals(0, sizeBuffer.getInt(3 * BaseRepeatedValueViewVector.SIZE_WIDTH));

      // check values
      assertEquals(12, ((BigIntVector) listViewVector.getDataVector()).get(0));
      assertEquals(-7, ((BigIntVector) listViewVector.getDataVector()).get(1));
      assertEquals(25, ((BigIntVector) listViewVector.getDataVector()).get(2));
      assertEquals(0, ((BigIntVector) listViewVector.getDataVector()).get(3));
      assertEquals(-127, ((BigIntVector) listViewVector.getDataVector()).get(4));
      assertEquals(127, ((BigIntVector) listViewVector.getDataVector()).get(5));
      assertEquals(50, ((BigIntVector) listViewVector.getDataVector()).get(6));

      assertEquals(3, listViewVector.getLastSet());
      listViewVector.validate();
    }
  }
```

Reply:
I think that makes more sense. (Possibly we should make you set the child vector first, then setOffset/setSize can check the values against the size of the child vector. Also, we should make sure to bounds check offset/size and make sure you don't set size to be negative etc.)

Reply:
> (Possibly we should make you set the child vector first, 

Yes this is better. 

> then setOffset/setSize can check the values against the size of the child vector. 

This is done by the `validate` function already. 

> Also, we should make sure to bounds check offset/size and 

I _think_ this is also part of the validate function? We could write an explicit one, but _not sure_ if that is very useful? 

> make sure you don't set size to be negative etc.)

Yes, validation at set.

Reply:
@lidavidm is the above approach okay?

Reply:
I'd like to have at least basic bounds checks like no negative values. It's better to fail early. We can always remove them if performance is a complaint later.

Reply:
For now, I added negative value validation.  Let me check if we can add the full validation.